File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
dp=[[-1 for _ in range(501)] for _ in range(501)]
def go(x,y,t):
    if y>=x+1 or x==len(t):
        return 0
    if dp[x][y]!=-1:
        return dp[x][y]
    ret=0
    ret+=max(ret,go(x+1,y+1,t),go(x+1,y,t))+t[x][y]
    dp[x][y]=ret
    return dp[x][y]
def solution(triangle):
    return go(0,0,triangle)

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.cc
#include <bits/stdc++.h>
using namespace std;
int dy[2]={0,1};
int dp[501][501];
int t[501][501];
int N;
int solve(int x,int y){
    if (x==N-1){
        return t[x][y];
    }
    int&ret=dp[x][y];
    if (ret!=-1)return ret;
    return ret=max(solve(x+1,y),solve(x+1,y+1))+t[x][y];
}
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
    cin >> N;
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j <= i; j++)
        {
            cin >> t[i][j];
        }
    }
    memset(dp, -1, sizeof(dp));
    cout << solve(0, 0);
}